src/libraries/System.Private.CoreLib/src/System/IO/BinaryReader.cs

```diff
@@ -504,40 +493,41 @@ private ReadOnlySpan<byte> InternalRead(int numBytes)
         // reasons. More about the subject in: https://github.com/dotnet/coreclr/pull/22102
         protected virtual void FillBuffer(int numBytes)
         {
-            if (numBytes < 0 || numBytes > _buffer.Length)
-            {
-                throw new ArgumentOutOfRangeException(nameof(numBytes), SR.ArgumentOutOfRange_BinaryReaderFillBuffer);
-            }
+            ArgumentOutOfRangeException.ThrowIfNegative(numBytes);
 
             ThrowIfDisposed();
 
-            // Need to find a good threshold for calling ReadByte() repeatedly
-            // vs. calling Read(byte[], int, int) for both buffered & unbuffered
-            // streams.
-            if (numBytes == 1)
-            {
-                int n = _stream.ReadByte();
-                if (n == -1)
-                {
-                    ThrowHelper.ThrowEndOfFileException();
-                }
-
-                _buffer[0] = (byte)n;
-                return;
-            }
-
-            if (numBytes > 0)
+            switch (numBytes)
             {
-                _stream.ReadExactly(_buffer.AsSpan(0, numBytes));
-            }
-            else
-            {
-                // ReadExactly no-ops for empty buffers, so special case numBytes == 0 to preserve existing behavior.
-                int n = _stream.Read(_buffer, 0, 0);
-                if (n == 0)
-                {
-                    ThrowHelper.ThrowEndOfFileException();
-                }
+                case 0:
+                    // ReadExactly no-ops for empty buffers, so special case numBytes == 0 to preserve existing behavior.
+                    int n = _stream.Read(Array.Empty<byte>(), 0, 0);
+                    if (n == 0)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    break;
+                case 1:
+                    // Need to find a good threshold for calling ReadByte() repeatedly
+                    // vs. calling Read(byte[], int, int) for both buffered & unbuffered
+                    // streams.
+                    n = _stream.ReadByte();
+                    if (n == -1)
+                    {
+                        ThrowHelper.ThrowEndOfFileException();
+                    }
+                    break;
+                default:
+                    byte[] buffer = ArrayPool<byte>.Shared.Rent(numBytes);
```

Comment:
In such case we can simply seek by `numBytes`-bytes?

```suggestion
                    if (_stream.CanSeek)
                    {
                        _stream.Seek(numBytes, SeekOrigin.Current);
                        return;
                    }

                    byte[] buffer = ArrayPool<byte>.Shared.Rent(numBytes);
```

Reply:
Nice idea, thanks! Fixed.